### ISSUE TITLE
[bug] graph generation fails when manipulating `cpp_info` from both `build_requirements` and `--profile:build` 

### ISSUE THREAD

**Issue by saterj** [conan-io/conan#6865]
### Environment Details (include every applicable attribute)
  * Operating System+version: Ubuntu 14.04 x86
  * Compiler+version: gcc4.9
  * Conan version: 1.24
  * Python version: 3.5

### Steps to reproduce (Include if Applicable)
1. Have a conan profile that manipulates a recipe's `cpp_info`
2. Have a recipe that has another project as it's `build_requirement` as it is [for this dependency on ](https://github.com/saterj/conan_codependent_lib/blob/master/conanfile.py) [this header-only project](https://github.com/saterj/header_only_conan)
3. attempt to create this new project with a `--profile:build` flag.  As with the [example's profile](https://github.com/saterj/conan_codependent_lib/tree/master/toolchain), I believe it has to be manipulating your `cpp_info`: `conan create 
4. The build should fail as your build graph doesn't have the generated dependencies (mine fails out because of the missing header)

### Logs (Executed commands with output) (Include/Attach if Applicable)
```conan create conan_codependent_lib Parts_Store/test --profile:build ECU400```
generated 
[test1.txt](https://github.com/conan-io/conan/files/4483611/test1.txt)

My profile is based on this:
```[settings]                     
os=Linux                       
os_target=Linux                
os_build=Linux                 
arch=x86                       
arch_target=armv7              
arch_build=armv7               
                               
[options]                      
[build_requires]               
*: ecu_sdk/2.7.2.220@SDKS/test
[env]                          
```

with that requires resolving to [this](https://github.com/saterj/conan_codependent_lib/blob/master/toolchain/conanfile.py)

**Comment by jgsogo** [comment]
Hi, @saterj 

Thanks for trying the new feature with the build/host contexts and for reporting back, we are still working on it, there are pending things and probably we will need to iterate it to fix issues from the user feedback.

First of all, this new feature is intended to supersede the old approach with the `os_build`, `arch_build`, `os_target` and `arch_target` settings. For Conan 2.0 if everything works as expected we will remove the old approach and after merging a couple of PRs for Conan 1.25 we will probably raise an error if there is a `---profile:build` and any of the settings mentioned before.

Said that, you need to choose if you go for the old approach or the new one. Mixing them will only generate troubles:
 * using the old approach there is only one context with all the settings from the `--profile`, each recipe will take into account the settings listed in the `conanfile.py` and that is the reason why `os_build` and `arch_build` exists, so applications running in the build_machine can take different values for these settings
 * using the new approach, with the `--profile:build` and the `--profile/--profile:host` there will be two different contexts, the `host` context will take the settings from the `--profile:host` and the build_context from the `--profile:build`, so every context have a full profile available with all the settings. There are also some details regarding `cpp_info` and `env_info` propagation: `cpp_info` propagates only in the host context, while `env_info` propagates from the build context to the host one.

Let me know the path you want to follow, share with us the `conan_codependent_lib` conanfile you are using and I'll try to help you.

Thanks!

**Comment by saterj** [comment]
The conanfile for the `conan_codependent_lib` is [here](https://github.com/saterj/conan_codependent_lib/blob/master/conanfile.py).

I think I'll try to build that project with the `--profile:host` set to the `ECU400` instead of `--profile:build`. It seems a little strange that the `cpp_info` only propogates for host profiles, but maybe I just got them mixed up. I would assume that build would describe the host that is currently building the project. 
TLDR I'll try your latter approach and remove the old way of doing it. This might've just been a silly question.

**Comment by jgsogo** [comment]
We use the terminology from GNU triplet convention:
 * build: the machine where the compilation is taking place
 * host: the machine where the generated binaries will run
 * target (only make sense from the POV of a compiler): the machine where files compiled with a compiler will run

In your graph, if `ECU400` is a compiler it has to be a `build_requires` in the `host` context (written as `self.build_requires(ECU)` in your recipe, or better in the `profile:host` file). If `HeaderOnly/trunk@Parts_Store/test` is a dependency of your project, it will never be a build requires but a `requires`.

**Comment by memsharded** [comment]
Closing as outdated and solved in Conan 2 with the new build/host model.

If there are any questions or issues, please open a new ticket against Conan 2, thanks for your feedback.
